File: booster-commons/src/main/java/io/github/booster/commons/retry/RetryConfig.java
package io.github.booster.commons.retry;

import io.github.booster.commons.metrics.MetricsRegistry;
import io.github.booster.commons.pool.NamedObjectPool;
import io.github.resilience4j.retry.Retry;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashMap;
import java.util.Map;

/**
 * Provides a central repository for {@link Retry} management.
 */
public class RetryConfig extends NamedObjectPool<Retry> {

    private static final Logger log = LoggerFactory.getLogger(RetryConfig.class);

    private Map<String, RetrySetting> settings;

    private MetricsRegistry registry;

    /**
     * Default constructor
     */
    public RetryConfig() {
        this(null);
    }

    /**
     * Constructor with default retry settings.
     * @param settings map of {@link RetrySetting} identified by name
     */
    public RetryConfig(Map<String, RetrySetting> settings) {
        this.settings = settings == null ? new HashMap<>() : settings;
    }

    public void setSettings(Map<String, RetrySetting> settings) {
        this.settings = settings == null ? new HashMap<>() : settings;
    }

    public void setMetricsRegistry(MetricsRegistry registry) {
        this.registry = registry == null ?
                new MetricsRegistry() :
                registry;
    }

    @Override
    protected Retry createObject(String name) {
        log.debug("booster-commons - cache contains [{}] entry: {}", name, this.settings.containsKey(name));
        return this.settings.containsKey(name) ?
                this.settings.get(name).buildRetry(name, this.registry).orNull() :
                null;
    }
}

File: booster-commons/src/main/kotlin/io/github/booster/commons/retry/RetryConfig.kt
package io.github.booster.commons.retry

import io.github.booster.commons.cache.GenericKeyedObjectCache
import io.github.booster.commons.cache.KeyedCacheObjectFactory
import io.github.booster.commons.cache.KeyedObjectCache
import io.github.booster.commons.metrics.MetricsRegistry
import io.github.resilience4j.retry.Retry
import org.slf4j.LoggerFactory

/**
 * Provides a central repository for [Retry] management.
 */
class RetryConfig @JvmOverloads constructor(settings: Map<String, RetrySetting>? = null) :
    KeyedCacheObjectFactory<String, Retry>, KeyedObjectCache<String, Retry> {

    private var settings: Map<String, RetrySetting>

    private var registry: MetricsRegistry? = null

    private var pool: GenericKeyedObjectCache<String, Retry>

    /**
     * Constructor with default retry settings.
     * @param settings map of [RetrySetting] identified by name
     */
    init {
        this.settings = settings ?: HashMap()
        this.pool = GenericKeyedObjectCache(this)
    }

    override fun create(key: String): Retry? {
        log.debug("booster-commons - cache contains [{}] entry: {}", key, settings.containsKey(key))
        return if (settings.containsKey(key))
            settings[key]!!.buildRetry(key, registry).orNull()
        else null
    }

    override fun get(key: String): Retry? = this.pool.get(key)

    fun setSettings(settings: Map<String, RetrySetting>?) {
        this.settings = settings ?: mapOf()
        this.pool = GenericKeyedObjectCache(this)
    }

    fun getSettings() = this.settings

    fun setMetricsRegistry(registry: MetricsRegistry?) {
        this.registry = registry ?: MetricsRegistry()
    }

    companion object {
        private val log = LoggerFactory.getLogger(RetryConfig::class.java)
    }

    override fun getKeys(): Set<String> = this.pool.getKeys()
}
